clients/command-kafka/src/main/java/org/eclipse/hono/client/command/kafka/KafkaBasedInternalCommandConsumer.java

```diff
@@ -321,20 +358,43 @@ private String getTopicName() {
         return new HonoTopic(HonoTopic.Type.COMMAND_INTERNAL, adapterInstanceId).toString();
     }
 
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Closes the Kafka admin client and consumer.
+     *
+     * @return A future indicating the outcome of the operation.
+     *         The future will be succeeded once this component is stopped.
+     */
     @Override
     public Future<Void> stop() {
-        retryCreateTopic.set(false);
-        vertx.cancelTimer(retryCreateTopicTimerId);
 
-        if (consumer == null) {
-            return Future.failedFuture("not started");
+        if (lifecycleStatus.isStopped()) {
+            // nothing to do
+            return Future.succeededFuture();
         }
 
-        return CompositeFuture.all(closeAdminClient(), closeConsumer())
-                .mapEmpty();
+        final Promise<Void> result = Promise.promise();
+        lifecycleStatus.addOnStoppedHandler(result);
+        if (lifecycleStatus.isStopping()) {
+            return result.future();
+        }
+
+        lifecycleStatus.setStopping();
```

Comment:
These lines are duplicated in 4 classes now.
How about putting them in a new `LifecycleStatus` method, returning the result Promise/Future?

Reply:
Let me see what I can do ...